zrml/swaps/fuzz/data_structs.rs

```diff
@@ -0,0 +1,109 @@
+#![allow(
+    // Mocks are only used for fuzzing and unit tests
+    clippy::integer_arithmetic
+)]
+
+use zeitgeist_primitives::constants::{MaxAssets, MaxTotalWeight, MinAssets, MinWeight};
+
+use arbitrary::{Arbitrary, Result, Unstructured};
+
+use rand::Rng;
+
+#[derive(Debug)]
+pub struct ValidPoolData {
+    pub origin: u128,
+    pub assets: Vec<(u128, u16)>,
+    pub base_asset: (u128, u16),
+    pub market_id: u128,
+    pub swap_fee: u128,
+    pub weights: Vec<u128>,
+}
+
+impl<'a> arbitrary::Arbitrary<'a> for ValidPoolData {
+    fn arbitrary(_: &mut Unstructured<'a>) -> Result<Self> {
+        let mut rng = rand::thread_rng();
+
+        let assets_len: u16 = rng.gen_range(MinAssets::get()..=MaxAssets::get());
+
+        let assets_len = assets_len as usize;
+        // unstructured arbitrary_len function did always return zero, that's why using rand crate
+        // create a weight collection with the capacity of assets length
+        let mut weights: Vec<u128> = Vec::with_capacity(assets_len);
+        let mut weight_sum = 0u128;
+
+        let mut assets: Vec<(u128, u16)> = Vec::with_capacity(assets_len);
+
+        for _ in 0..assets_len {
+            // use always the min weight to rarely reach the constraint of MaxTotalWeight
+            let weight: u128 = MinWeight::get();
+            match weight_sum.checked_add(weight) {
+                // MaxWeight is 50 * BASE and MaxTotalWeight is also 50 * BASE
+                // so it is very likely to reach the max total weight with two assets
+                Some(sum) if sum <= MaxTotalWeight::get() => weight_sum = sum,
+                // if sum > MaxTotalWeight or u128 Overflow (None case)
+                _ => return Err(<arbitrary::Error>::IncorrectFormat),
+            }
+            let asset = (rng.gen::<u128>(), rng.gen::<u16>());
+
+            weights.push(weight);
+            assets.push(asset);
+        }
```

Comment:
The struct ValidPoolData is only used to create a random valid pool. 
I agree, sure, the `weight` is not as random as it could be, but for constructing a valid pool, as I used a random weight, the `MaxTotalWeight` was reached very often and therefore the pool was not created. It could be modified to the following, if this is wanted.

If the `match` is to complicated I could reform it with `if` statements. But I like the style, because it is smaller than the other one.

I don't really get your second bullet point. But I would modify the weight anyways like the following.

```rust
let weight: u128 = rand::thread_rng().gen_range(MinWeight ..= MaxTotalWeight / assets_len as u128);
```
The above is only to ensure, that the sum of weights doesn't reach the `MaxTotalWeight`.
Is this acceptable?

Reply:
Is there a quick rust function, that divides a sum into random (but at least `MinWeight`) parts?

```rust 
let mut target_total_weight = rng.gen_range(MinWeight::get()*assets_len as u128..=MaxTotalWeight::get());
```

The above is the total and random weight sum, which would create a valid pool.

Reply:
Yeah, the second bullet point wasn't very clear. What I meant was: `assets_len` is at most `MaxAssets`, which is 12. So if you take `weight` to always be `MinWeight` (which is `BASE`), then you never exceed `MaxTotalWeight` (which is `50 * BASE`). So there's no need to check if the `MaxTotalWeight` is exceeded.

I don't think there's a builtin function that splits a sum into random parts, but that should be easy to write. Something like this (please double check, this is pure improv):

```
fn create_random_weights(count: u16) -> Vec<u128> {
    let result = vec![];
    for i in 0..count {
        let max =
            (T::MaxTotalWeight::get() - (count - i) * T::MinWeight::get()).max(T::MaxWeight::get());
        result.push(rng.gen_range(T::MinWeight::get()..max));
    }
    // Need to shuffle the vector, because earlier numbers have a higher probability of being
    // large.
    vec.shuffle(&mut thread_rng()); 
    result
}
```

Anyways, having the same weight for all assets is a bit too non-random, I think.

Reply:
I understand, that it's a bit to non-random at the moment. Yeah you are right, that checking for MaxTotalWeight is unnecessary in my case. 

To your code: If you decide the maximum between `MaxWeight` and the other:
`MaxWeight == 50 * BASE` and `MaxTotalWeight == 50 * BASE`

Then if you randomly select in range `MinWeight` to `50 * BASE` you'll get a nearly 50% chance of getting already one weight of **0.5** * `50 * BASE` and therefore a high probability to reach the MaxTotalWeight with the second push to the vector. 

The implementation of this seems more complicated. @maltekliemann @sea212 Feel free to propose another algorithm for this.

Reply:
Oh yeah, my bad. I meant

```rust
fn create_random_weights(count: u16) -> Vec<u128> {
    let result = vec![];
    let sum = 0;
    for i in 0..count {
        let max =
            (T::MaxTotalWeight::get() - (count - i) * T::MinWeight::get()).max(T::MaxWeight::get());
        let weight = rng.gen_range(T::MinWeight::get()..max);
        sum += weight;
        result.push(weight);
    }
    // Need to shuffle the vector, because earlier numbers have a higher probability of being
    // large.
    vec.shuffle(&mut thread_rng()); 
    result
}
```

Reply:
This looks well thought to me. :-) I am nearly happy with that implementation now. If I assume, that `MaxWeight == MaxTotalWeight` stays, then it could be:

```rust
let max = MaxWeight::get() - sum;
```

Reply:
But without using assumptions yours is more concrete. I will use it, thanks!

Reply:
Cool. I think it may have to be `count - 1 - i` instead of `count - i`, though, right?

Reply:
We only need to make sure, that `MaxWeight` is not larger than `TotalMaxWeight` in the future. Otherwise there will be intensive debugging, why the fuzz tests don't create a valid pool.

Reply:
```rust 
if MaxWeight::get() > MaxTotalWeight::get() {
    panic!("Unexpected error during random weight creation. MaxWeight should be smaller than MaxTotalWeight");
}
```

Reply:
For the first round there are `(assets_len - 1)`assets left, which need to be reserved, right.

```rust
let max = (MaxTotalWeight::get() - (count - 1 - i) * MinWeight::get()).max(MaxWeight::gett()) - sum;
```

Reply:
Mhhh, but why choosing between the both Max(Total)Weights at all, when `MaxWeight == MaxTotalWeight` and it would always be `MaxWeight` with this implementation.. I think, that throwing a `panic!` if MaxWeight is greater MaxTotalWeight should be required here.

```rust 
    if MaxWeight::get() > MaxTotalWeight::get() {
        panic!(
            "Unexpected error during random weight creation. MaxWeight should be smaller than \
             MaxTotalWeight."
        );
    }

let max = (MaxTotalWeight::get() - (count - 1 - i) * MinWeight::get()) - sum;
```

Reply:
That's true, but as far as I'm concerned, you don't need that check. `MaxTotalWeight` is a maximum on the sum of all weights, while `MaxWeight` is the maximum for each individual weight. Configuring the pallet to have `MaxWeight < MaxTotalWeight` would be very weird, and I don't think we will ever modify the current configuration.

Reply:
You mean it's weird to have `MaxWeight > MaxTotalWeight` ? I agree.

```rust
        // first iteration: (asset_len - 1) assets left
        let assets_left = assets_len - 1 - i;
        // reservation of multiple MinWeight for the future iterations
        let future_min_weight_reserve = assets_left * MinWeight::get();
        // take min_weight_reserve for future iterations in calculation
        // maximum value of weight without looking at the previous weight_sum
        let max_weight_limit = MaxTotalWeight::get() - future_min_weight_reserve;
        // previous weight sum substraction limits to exceed the MaxTotalWeight (otherwise no pool creation)
        let max_weight_limit = max_weight_limit - weight_sum;
        // each individual weight is at most MaxWeight
        let max = max_weight_limit.min(MaxWeight::get());
        let weight = rng.gen_range(MinWeight::get()..max);
```

Reply:
Yeah, `MaxWeight > MaxTotalWeight`, not the other way around 🙈

Reply:
Please @sea212, take a deeper look into this function.